wagtail/admin/views/collections.py

```diff
@@ -140,12 +140,8 @@ def save_instance(self):
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
-        context["can_delete"] = (
-            self.permission_policy.instances_user_has_permission_for(
-                self.request.user, "delete"
-            )
-            .filter(pk=self.object.pk)
-            .first()
+        context["can_delete"] = self.permission_policy.user_has_permission_for_instance(
```

Comment:
A different way to solve this would be to pass `can_delete` as a kwarg to `super().get_context_data()`, so that they are added to the context before the rest of the generic `EditView.get_context_data()` runs. Then, in the generic `EditView`, we'll need to change this line to only set the `can_delete` value if it is not already in the context:

https://github.com/wagtail/wagtail/blob/bf3f87b75910495471e84e976fff0af7f6a74df5/wagtail/admin/views/generic/models.py#L920-L923

However, this means all subclasses that override `can_delete` need to be updated to use this pattern. It's not a big deal though, as within Wagtail itself, the only other `EditView` subclass that overrides `can_delete` I can find is the user edit view. Even so, it does not have this issue as it uses a custom template.

It overrides the `can_delete` here:

https://github.com/wagtail/wagtail/blob/bf3f87b75910495471e84e976fff0af7f6a74df5/wagtail/users/views/users.py#L330-L333

Using the following `user_can_delete_user` util (which by the way, could be implemented via a `ModelPermissionTester` or `DeleteAction` subclass if we had one...)

https://github.com/wagtail/wagtail/blob/bf3f87b75910495471e84e976fff0af7f6a74df5/wagtail/users/views/users.py#L291-L295

https://github.com/wagtail/wagtail/blob/bf3f87b75910495471e84e976fff0af7f6a74df5/wagtail/users/utils.py#L13-L25

But the view uses a custom template that checks `can_delete`, so it doesn't have the same issue:

https://github.com/wagtail/wagtail/blob/bf3f87b75910495471e84e976fff0af7f6a74df5/wagtail/users/templates/wagtailusers/users/edit.html#L70-L72

Although, I suppose the template can use a bit more refactoring e.g. using `delete_url` instead of `{% url %}`... I considered doing this in 4d305d24203705225c7499bd17d5dc7cd9b74c24, but ultimately decided not to, to avoid unnecessary risks and to make the PR more focused on just introducing the `UserViewSet`. I'd be happy to do it in a separate PR, though!


I also found a few other view subclasses that have `can_{foo}` context, but from what I've seen they seem to add these context values rather than modify existing ones. For example:

- `can_delete` in locale delete view
  - The generic delete view does not have this context variable. The locale delete view uses this to render a message when deleting the locale is not allowed https://github.com/wagtail/wagtail/blob/bf3f87b75910495471e84e976fff0af7f6a74df5/wagtail/locales/views.py#L68-L86 https://github.com/wagtail/wagtail/blob/bf3f87b75910495471e84e976fff0af7f6a74df5/wagtail/locales/templates/wagtaillocales/confirm_delete.html#L5-L13
  - Note that the message is rendered in the template. This means the view will load normally (with 200 status code) instead of a redirect with that message displayed as an error banner at the top like we usually do when the user does not have enough permissions.
  - Also, again, this additional logic could be implemented via a `ModelPermissionTester` or `DeleteAction` subclass if we had one...
- `can_enable`/`can_disable` in workflows and tasks edit view
  - These are domain-specific, since you can't delete workflows and tasks. You can only disable them. The templates make use of these additional variables.